File: Subsets II/Subsets II.java
public class Solution {
    
    private void recur(List<List<Integer>> ret, List<Integer> solution, int[] num, int level){
        int size = num.length;
        
        for(int i = level; i < size; ++i){
            solution.add(num[i]);
            ret.add(new ArrayList<Integer>(solution));
            if(i != (size - 1)){
                recur(ret, solution, num, i + 1);
                while(num[i] == num[i + 1]){
                    ++i;
                    if(i == (size - 1)){
                        break;
                    }
                }
            }
            solution.remove(solution.size() - 1);
        }
        
        return;
    }
    
    public List<List<Integer>> subsetsWithDup(int[] num) {
        
        List<Integer> set = new ArrayList<Integer>();
        List<List<Integer>> ret = new ArrayList<List<Integer>>();
        
        ret.add(set);
        Arrays.sort(num);
        recur(ret, set, num, 0);
        
        return ret;
    }
}

File: Subsets II/Subsets II.cpp
class Solution {
public:

    void helper(std::vector<std::vector<int> >& res, std::vector<int>& currSol, int index, std::vector<int>& S){
        for(int i = index; i < S.size(); ++i){
            currSol.push_back(S[i]);
            res.push_back(currSol);
            if(i < S.size() - 1){
                helper(res, currSol, index + 1, S);
                while(i < S.size() - 1 && S[i] == S[i + 1])     ++i;
            }
            currSol.pop_back();
        }
    }

    vector<vector<int> > subsetsWithDup(vector<int> &S) {
        std::vector<std::vector<int> > res;
        std::vector<int> sol;
        //std::sort(S.begin(), S.end(), [](int a, int b){return a < b;});
        std::sort(S.begin(), S.end());
        res.push_back(sol);
        helper(res, sol, 0, S);
        return res;
    }
};
